File: extractapi-linux-x86_64-2019_2/share/tableausdk-2019.2.6.199.r40e5865b/samples/TableauSDKSample.cpp
//------------------------------------------------------------------------------
//
//  This file is the copyrighted property of Tableau Software and is protected
//  by registered patents and other applicable U.S. and international laws and
//  regulations.
//
//  Unlicensed use of the contents of this file is prohibited. Please refer to
//  the NOTICES.txt file for further details.
//
//------------------------------------------------------------------------------
#if defined(__APPLE__) && defined(__MACH__)
#include <TableauHyperExtract/TableauHyperExtract_cpp.h>
#else
#include "TableauHyperExtract_cpp.h"
#endif

#include <codecvt>
#include <cstring>
#include <iostream>
#include <locale>
#include <map>
#include <string>

using namespace Tableau;

//------------------------------------------------------------------------------
//  Display Usage
//------------------------------------------------------------------------------
void DisplayUsage()
{
    std::cerr << "A simple demonstration of the Tableau Extract API:" << std::endl
              << std::endl
              << "USAGE: tableauSDKSample-cpp [COMMAND] [OPTIONS]" << std::endl
              << std::endl
              << "COMMANDS:" << std::endl
              << "  -h, --help           Show this help message and exit" << std::endl
              << std::endl
              << "  -b, --build          If a Tableau extract named FILENAME exists in the current directory,"
              << std::endl
              << "                       extend it with sample data." << std::endl
              << "                       If no Tableau extract named FILENAME exists in the current directory,"
              << std::endl
              << "                       create one and populate it with sample data." << std::endl
              << std::endl
              << "OPTIONS:" << std::endl
              << " -s, --spatial         Include spatial data when creating a new extract." << std::endl
              << "                       If an extract is being extended, this argument is ignored." << std::endl
              << "                       (default=False)" << std::endl
              << std::endl
              << " -m, --multitable      Create multi-table extract." << std::endl
              << "                       Note that multi-table storage for extracts is supported starting with Tableau 2018.3." << std::endl
              << "                       (default=False)" << std::endl
              << std::endl
              << " -f FILENAME, --filename FILENAME" << std::endl
              << "                       FILENAME of the extract to be created or extended." << std::endl
              << "                       (default='order-cpp.hyper')" << std::endl;
}

//------------------------------------------------------------------------------
//  Parse Arguments
//------------------------------------------------------------------------------
//  Parse Command Line Arguments or Populate Defaults
//
//  Returns 'true' if all arguments are successfully parsed
//  Returns 'false' if there are any invalid arguments or no arguments
bool ParseArguments(int argc, char* argv[], std::map<std::string, std::wstring>& options)
{
    if (argc == 0)
    {
        return false;
    }

    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t, 0x10ffff, std::codecvt_mode::little_endian>> converter;
    for (int i = 0; i < argc; ++i)
    {
        if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help"))
        {
            options[std::string("help")] = std::wstring(L"true");
            return true;
        }
        else if (!strcmp(argv[i], "-b") || !strcmp(argv[i], "--build"))
        {
            options[std::string("build")] = std::wstring(L"true");
        }
        else if (!strcmp(argv[i], "-s") || !strcmp(argv[i], "--spatial"))
        {
            options[std::string("spatial")] = std::wstring(L"true");
        }
        else if (!strcmp(argv[i], "-m") || !strcmp(argv[i], "--multitable"))
        {
            options[std::string("multitable")] = std::wstring(L"true");
        }
        else if (!strcmp(argv[i], "-f") || !strcmp(argv[i], "--filename"))
        {
            if (i >= argc)
            {
                return false;
            }
            options[std::string("filename")] = converter.from_bytes(argv[++i]);
        }
        else
        {
            return false;
        }
    }

    //  Defaults
    //  options["build"] = "false";
    //  options["spatial"] = "false";
    if (!options.count("filename"))
    {
        options["filename"] = L"order-cpp.hyper";
    }

    return true;
}

//------------------------------------------------------------------------------
//  Create or Open Extract
//------------------------------------------------------------------------------
std::shared_ptr<Extract> CreateExtract(const std::wstring& fileName, const bool useSpatial, bool createMultipleTables)
{
    std::shared_ptr<Extract> extractPtr = nullptr;
    try
    {
        //  Create Extract Object
        //  (NOTE: TabExtractCreate() opens an existing extract with the given
        //   filename if one exists or creates a new extract with the given filename
        //   if one does not)
        extractPtr = std::make_shared<Extract>(fileName);

        //  Define `Extract`/`Products` table
        auto tableName = createMultipleTables ? L"Products" : L"Extract";
        if (!extractPtr->HasTable(tableName))
        {
            TableDefinition schema;
            // NOTE: Collations make string operations very expensive and this may
            // slow down the overall workbook performance, so use them with care.
            schema.SetDefaultCollation(Collation_Binary);
            schema.AddColumn(L"Purchased", Type_DateTime);
            schema.AddColumn(L"Product", Type_CharString);
            schema.AddColumn(L"uProduct", Type_UnicodeString);
            schema.AddColumn(L"Price", Type_Double);
            schema.AddColumn(L"Quantity", Type_Integer);
            schema.AddColumn(L"Taxed", Type_Boolean);
            schema.AddColumn(L"Expiration Date", Type_Date);
            schema.AddColumnWithCollation(L"Produkt", Type_CharString, Collation_de);
            if (useSpatial)
            {
                schema.AddColumn(L"Destination", Type_Spatial);
            }
            if (createMultipleTables)
            {
                schema.AddColumn(L"SupplierKey", Type_Integer);
            }

            std::shared_ptr<Table> tablePtr = extractPtr->AddTable(tableName, schema);
            if (tablePtr == nullptr)
            {
                std::wcerr << L"A fatal error occurred while creating the new table" << std::endl
                           << L"Exiting Now." << std::endl;
                exit(EXIT_FAILURE);
            }
        }

        //  Define `Suppliers` table (If we are creating a new multi-table extract)
        if (createMultipleTables && !extractPtr->HasTable(L"Suppliers"))
        {
            TableDefinition schema;
            schema.AddColumn(L"SupplierKey", Type_Integer);
            schema.AddColumn(L"Supplier", Type_CharString);
            schema.AddColumn(L"Address", Type_CharString);

            std::shared_ptr<Table> tablePtr = extractPtr->AddTable(L"Suppliers", schema);
            if (tablePtr == nullptr)
            {
                std::wcerr << L"A fatal error occurred while creating the `Suppliers` table" << std::endl
                           << L"Exiting Now." << std::endl;
                exit(EXIT_FAILURE);
            }
        }
    }
    catch (const TableauException& e)
    {
        std::wcerr << L"A fatal error occurred while creating the new extract: " << std::endl
                   << e.GetMessage() << std::endl
                   << L"Exiting Now." << std::endl;
        exit(EXIT_FAILURE);
    }

    return extractPtr;
}

//------------------------------------------------------------------------------
//  Populate Extract
//------------------------------------------------------------------------------
void PopulateExtract(const std::shared_ptr<Extract> extractPtr, bool useSpatial, bool createMultipleTables)
{
    try
    {
        // Populate `Extract`/`Products` table
        {
            //  Get Schema
            auto tableName = createMultipleTables ? L"Products" : L"Extract";
            std::shared_ptr<Table> tablePtr = extractPtr->OpenTable(tableName);
            std::shared_ptr<Tableau::TableDefinition> schema = tablePtr->GetTableDefinition();

            //  Insert Data
            Tableau::Row row(*schema);
            row.SetDateTime(0, 2012, 7, 3, 11, 40, 12, 4550); //  Purchased
            row.SetCharString(1, "Beans");                    //  Product
            row.SetString(2, L"uniBeans");                    //  Unicode Product
            row.SetDouble(3, 1.08);                           //  Price
            row.SetDate(6, 2029, 1, 1);                       //  Expiration Date
            row.SetCharString(7, "Bohnen");                   //  Produkt
            if (useSpatial)
            {
                row.SetSpatial(8, "POINT (30 10)"); //  Destination
            }
            for (int i = 0; i < 10; ++i)
            {
                row.SetInteger(4, i * 10);     //  Quantity
                row.SetBoolean(5, i % 2 == 1); //  Taxed
                if (createMultipleTables)
                {
                    row.SetInteger(useSpatial ? 9 : 8, i % 3); //  SupplierKey
                }
                tablePtr->Insert(row);
            }
        }

        // Populate `Suppliers` table
        if (createMultipleTables) {
            //  Get Schema
            std::shared_ptr<Table> tablePtr = extractPtr->OpenTable(L"Suppliers");
            std::shared_ptr<Tableau::TableDefinition> schema = tablePtr->GetTableDefinition();

            //  Insert Data
            Tableau::Row row(*schema);
            row.SetCharString(1, "Bean Supplier");            //  Supplier
            row.SetCharString(2, "42 Bean Street, Beantown"); //  Address
            for (int i = 0; i < 3; ++i)
            {
                row.SetInteger(0, i);                         //  SupplierKey
                tablePtr->Insert(row);
            }
        }
    }
    catch (const TableauException& e)
    {
        std::wcerr << L"A fatal error occurred while populating the extract: " << std::endl
                   << e.GetMessage() << std::endl
                   << L"Exiting Now." << std::endl;
        exit(EXIT_FAILURE);
    }
}

//------------------------------------------------------------------------------
//  Main
//------------------------------------------------------------------------------
int main(int argc, char* argv[])
{
    //  Parse Arguments
    std::map<std::string, std::wstring> options;
    if (!ParseArguments(argc - 1, argv + 1, options) || options.count("help"))
    {
        DisplayUsage();
        exit(EXIT_SUCCESS);
    }

    //  Extract API Demo
    if (options.count("build") > 0)
    {
        //  Initialize the Tableau Extract API
        ExtractAPI::Initialize();

        //  Create or Expand the Extract
        std::shared_ptr<Extract> extractPtr = nullptr;
        extractPtr = std::move(CreateExtract(options["filename"], options.count("spatial") > 0, options.count("multitable") > 0));
        PopulateExtract(extractPtr, options.count("spatial") > 0, options.count("multitable") > 0);

        //  Flush the Extract to Disk
        extractPtr->Close();

        //  Close the Tableau Extract API
        ExtractAPI::Cleanup();
    }

    exit(EXIT_SUCCESS);
}

File: extractapi-linux-x86_64-2019_2/share/tableausdk-2019.2.6.199.r40e5865b/samples/com/tableausoftware/demos/TableauSDKSample.java
/**
 * -----------------------------------------------------------------------------
 *
 *  This file is the copyrighted property of Tableau Software and is protected
 *  by registered patents and other applicable U.S. and international laws and
 *  regulations.
 *
 *  Unlicensed use of the contents of this file is prohibited. Please refer to
 *  the NOTICES.txt file for further details.
 *
 * -----------------------------------------------------------------------------
 */
package com.tableausoftware.demos;

import java.util.HashMap;

import com.tableausoftware.TableauException;
import com.tableausoftware.common.*;
import com.tableausoftware.hyperextract.*;

public final class TableauSDKSample {
    /**
     * -------------------------------------------------------------------------
     *  Display Usage
     * -------------------------------------------------------------------------
     */
    private static void displayUsage() {
        System.err.print(""
            + "A simple demonstration of the Tableau Extract API.\n"
            + "\n"
            + "USAGE: tableauSDKSample-c [COMMAND] [OPTIONS]\n"
            + "\n"
            + "COMMANDS:"
            + "\n"
            + " -h, --help            Show this help message and exit\n"
            + " -b, --build           If an extract named FILENAME exists in the current directory,\n"
            + "                       extend it with sample data.\n"
            + "                       If no Tableau extract named FILENAME exists in the current directory,\n"
            + "                       create one and populate it with sample data.\n"
            + "                       (default=False)\n"
            + "\n"
            + "OPTIONS:\n"
            + " -s, --spatial         Include spatial data when creating a new extract.\n"
            + "                       If an extract is being extended, this argument is ignored.\n"
            + "                       (default=False)\n"
            + "\n"
            + " -m, --multitable      Create multi-table extract.\n"
            + "                       Note that multi-table storage for extracts is supported starting with Tableau 2018.3.\n"
            + "                       (default=False)\n"
            + "\n"
            + " -f FILENAME, --filename FILENAME\n"
            + "                       FILENAME of the extract to be created or extended.\n"
            + "                       (default='order-c.hyper')\n"
            + "\n" );
    }


    /**
     * -------------------------------------------------------------------------
     *  Parse Arguments
     * -------------------------------------------------------------------------
     *  (NOTE: This function returns 'null' if there are any invalid arguments)
     */
    private static HashMap<String, String> parseArguments(
        String[] arguments
    )
    {
        if ( arguments.length == 0 ) {
            return null;
        }

        HashMap<String, String> options = new HashMap<String, String>();
        for ( int i = 0; i < arguments.length; ++i ) {
            switch( arguments[ i ] ) {
            case "-h":
            case "--help":
                options.put( "help", "true" );
                return options;
            case "-b":
            case "--build":
                options.put( "build", "true" );
                break;
            case "-s":
            case "--spatial":
                options.put( "spatial", "true" );
                break;
            case "-m":
            case "--multitable":
                options.put( "multitable", "true" );
                break;
            case "-f":
            case "--filename":
                if ( i + 1 >= arguments.length ) {
                    return null;
                }
                options.put( "filename", arguments[ ++i ] );
                break;
            default:
                return null;
            }
        }

        //  Defaults
        //  options.put( "build" ) = false;
        //  options.put( "spatial" ) = false;
        if ( !options.containsKey( "filename" ) ) {
            options.put( "filename", "order-java.hyper" );
        }

        return options;
    }

    /**
     * -------------------------------------------------------------------------
     *  Create or Open Extract
     * -------------------------------------------------------------------------
     *  (NOTE: This function assumes that the Tableau SDK Extract API is initialized)
     */
    private static Extract createOrOpenExtract(
        String filename,
        boolean useSpatial,
        boolean createMultipleTables
    )
    {
        Extract extract = null;
        try {
            //  Create Extract Object
            //  (NOTE: TabExtractCreate() opens an existing extract with the given
            //   filename if one exists or creates a new extract with the given filename
            //   if one does not)
            extract = new Extract( filename );

            //  Define `Extract`/`Products` table (If we are creating a new extract)
            String tableName = createMultipleTables ? "Products" : "Extract";
            if ( !extract.hasTable( tableName ) ) {
                TableDefinition schema = new TableDefinition();
                // NOTE: Collations make string operations very expensive and this may
                // slow down the overall workbook performance, so use them with care.
                schema.setDefaultCollation( Collation.BINARY );
                schema.addColumn( "Purchased",       Type.DATETIME );
                schema.addColumn( "Product",         Type.CHAR_STRING );
                schema.addColumn( "uProduct",        Type.UNICODE_STRING );
                schema.addColumn( "Price",           Type.DOUBLE );
                schema.addColumn( "Quantity",        Type.INTEGER );
                schema.addColumn( "Taxed",           Type.BOOLEAN );
                schema.addColumn( "Expiration Date", Type.DATE );
                schema.addColumnWithCollation( "Produkt", Type.CHAR_STRING, Collation.DE );
                if ( useSpatial ) {
                    schema.addColumn( "Destination", Type.SPATIAL );
                }
                if ( createMultipleTables ) {
                    schema.addColumn( "SupplierKey",     Type.INTEGER );
                }
                Table table = extract.addTable( tableName, schema );
                if ( table == null ) {
                    System.err.println( "A fatal error occured while creating the table" );
                    System.err.println( "Exiting now." );
                    System.exit( -1 );
                }
            }

            //  Define `Suppliers` table (If we are creating a new multi-table extract)
            if ( createMultipleTables && !extract.hasTable( "Suppliers" ) ) {
                TableDefinition schema = new TableDefinition();
                schema.addColumn( "SupplierKey",     Type.INTEGER );
                schema.addColumn( "Supplier",        Type.CHAR_STRING );
                schema.addColumn( "Address",         Type.CHAR_STRING );
                Table table = extract.addTable( "Suppliers", schema );
                if ( table == null ) {
                    System.err.println( "A fatal error occured while creating the `Suppliers` table" );
                    System.err.println( "Exiting now." );
                    System.exit( -1 );
                }
            }
        }
        catch ( TableauException e ) {
            System.err.println( "A fatal error occurred while creating the extract:" );
            System.err.println( e.getMessage() );
            System.err.println( "Printing stack trace now:" );
            e.printStackTrace( System.err );
            System.err.println( "Exiting now." );
            System.exit( -1 );
        }
        catch ( Throwable t ) {
            System.err.println( "An unknown error occured while creating the extract" );
            System.err.println( "Printing stack trace now:" );
            t.printStackTrace( System.err );
            System.err.println( "Exiting now." );
            System.exit( -1 );
        }

        return extract;
    }

    /**
     * -------------------------------------------------------------------------
     *  Populate Extract
     * -------------------------------------------------------------------------
     *  (NOTE: This function assumes that the Tableau SDK Extract API is initialized)
     */
    private static void populateExtract(
        Extract extract,
        boolean useSpatial,
        boolean createMultipleTables
    )
    {
        try {
            // Populate `Extract`/`Products` table
            {
                //  Get Schema
                String tableName = createMultipleTables ? "Products" : "Extract";
                Table table = extract.openTable( tableName );
                TableDefinition tableDef = table.getTableDefinition();

                //  Insert Data
                Row row = new Row( tableDef );
                row.setDateTime( 0, 2012, 7, 3, 11, 40, 12, 4550 ); //  Purchased
                row.setCharString( 1, "Beans" );                    //  Product
                row.setString( 2, "uniBeans" );                     //  Unicode Product
                row.setDouble( 3, 1.08 );                           //  Price
                row.setDate( 6, 2029, 1, 1 );                       //  Expiration Date
                row.setCharString( 7, "Bohnen" );                   //  Produkt
                if ( useSpatial ) {
                    row.setSpatial( 8, "POINT (30 10)" );           //  Destination
                }
                for ( int i = 0; i < 10; ++i  ) {
                    row.setInteger( 4, i * 10 );                    //  Quantity
                    row.setBoolean( 5, i % 2 == 1 );                //  Taxed
                    if ( createMultipleTables ) {
                        row.setInteger( useSpatial ? 9 : 8, i % 3 ); //  SupplierKey
                    }
                    table.insert( row );
                }
            }

            // Populate `Suppliers` table
            if (createMultipleTables) {
                //  Get Schema
                Table table = extract.openTable( "Suppliers" );
                TableDefinition tableDef = table.getTableDefinition();

                //  Insert Data
                Row row = new Row( tableDef );
                row.setCharString( 1, "Bean Supplier" );            //  Supplier
                row.setCharString( 2, "42 Bean Street, Beantown" ); //  Address
                for ( int i = 0; i < 3; ++i  ) {
                    row.setInteger( 0, i );                         //  SupplierKey
                    table.insert( row );
                }
            }
        }
        catch ( TableauException e ) {
            System.err.println( "A fatal error occurred while populating the extract:" );
            System.err.println( e.getMessage() );
            System.err.println( "Printing stack trace now:" );
            e.printStackTrace( System.err );
            System.err.println( "Exiting now." );
            System.exit( -1 );
            }
        catch ( Throwable t ) {
            System.err.println( "An unknown error occured while populating the extract" );
            System.err.println( "Printing stack trace now:" );
            t.printStackTrace( System.err );
            System.err.println( "Exiting now." );
            System.exit( -1 );
        }
    }

    /**
     * -------------------------------------------------------------------------
     *  Main
     * -------------------------------------------------------------------------
     */
    public static void main( String[] arguments ) {
        //  Parse Arguments
        HashMap<String, String> options = parseArguments( arguments );
        if ( options == null || options.containsKey( "help" ) ) {
            displayUsage();
            return;
        }

        //  Extract API Demo
        if ( options.containsKey( "build" ) ) {
            try {
                //  Initialize the Tableau Extract API
                ExtractAPI.initialize();

                //  Create or Expand the Extract
                Extract extract = createOrOpenExtract( options.get( "filename" ), options.containsKey( "spatial" ), options.containsKey( "multitable" ) );
                populateExtract( extract, options.containsKey( "spatial" ), options.containsKey( "multitable" ) );

                //  Flush the Extract to Disk
                extract.close();

                // Close the Tableau Extract API
                ExtractAPI.cleanup();
            }
            catch ( TableauException e ) {
                System.err.println( "A fatal error occurred while opening or closing the Extract API:" );
                System.err.println( e.getMessage() );
                System.err.println( "Printing stack trace now:" );
                e.printStackTrace( System.err );
                System.err.println( "Exiting now." );
                System.exit( -1 );
            }
            catch ( Throwable t ) {
                System.err.println( "An unknown error occured while opening or closing the Extract API:" );
                System.err.println( "Printing stack trace now:" );
                t.printStackTrace( System.err );
                System.err.println( "Exiting now." );
                System.exit( -1 );
            }
        }
    }
}
